src/rust/engine/task_executor/src/lib.rs

```diff
@@ -200,4 +205,112 @@ impl Executor {
       workunit_store::scope_task_workunit_store_handle(workunit_store_handle, future).await
     })
   }
+
+  /// Return a reference to this executor's runtime handle.
+  pub fn handle(&self) -> &Handle {
+    &self.handle
+  }
+}
+
+/// Store "tail" tasks which are async tasks that can execute concurrently with regular
+/// build actions. Tail task block completion of a session until all of them have been
+/// completed (subject to a timeout).
+#[derive(Clone)]
+pub struct TailTasks {
+  inner: Arc<Mutex<Option<TailTasksInner>>>,
+}
+
+struct TailTasksInner {
+  id_to_name: HashMap<Id, String>,
+  task_set: JoinSet<()>,
+}
+
+impl TailTasks {
+  pub fn new() -> Self {
+    Self {
+      inner: Arc::new(Mutex::new(Some(TailTasksInner {
+        id_to_name: HashMap::new(),
+        task_set: JoinSet::new(),
+      }))),
+    }
+  }
+
+  /// Spawn a tail task with the given name.
+  pub fn spawn_on<F>(&self, name: &str, task: F, handle: &Handle)
+  where
+    F: Future<Output = ()>,
+    F: Send + 'static,
+  {
+    let mut guard = self.inner.lock();
+    let inner = match &mut *guard {
+      Some(inner) => inner,
+      None => {
+        log::warn!(
+          "Session end task `{}` submitted after session completed.",
+          name
+        );
+        return;
+      }
+    };
+
+    let h = inner.task_set.spawn_on(task, handle);
+    inner.id_to_name.insert(h.id(), name.to_string());
+  }
+
+  /// Wait for all tail tasks to complete subject to the given timeout. If tasks
+  /// fail or do not complete, log that fact.
+  pub async fn wait(self, timeout: Duration) {
+    let mut inner = match self.inner.lock().take() {
+      Some(inner) => inner,
+      None => {
+        log::debug!("Session end tasks awaited multiple times!");
```

Comment:
cc @stuhood

Reply:
`wait` is being called inside of `Scheduler.execute`, which is used multiple times within a single `Session`. For example: if you run multiple goals sequentially, it is executed as a loop over calls to `Scheduler.execute`.

If you wanted to only `wait` at the "end" of a `Session` (rather than once per execute call), you'd need to do something more like https://github.com/pantsbuild/pants/pull/16952#discussion_r977941836